src/components/annual-report/report-data/section2/section2dCalculations.js

```diff
@@ -0,0 +1,158 @@
+import { convertAnalyticsResponseToObject, getVal } from '../utils/utils.js'
+import { getForecastValue, getMean, getRoundedValue } from './mathService.js'
+import {
+    OVERALL_ORG_UNIT_SECTION_2D,
+    LEVEL_OR_GROUP_SECTION_2D,
+} from './useFetchSectionTwoData.js'
+
+const get2dScore = ({
+    ou,
+    dx,
+    trend,
+    response,
+    currentPeriodID,
+    comparisonPeriods,
+}) => {
+    const currentPeriodValue = Number(
+        getVal({ response, dx, pe: currentPeriodID, ou })
+    )
+
+    // need to keep relative distance for forecast (hence get index, then filter out invalid values)
+    const comparisonPeriodPoints = comparisonPeriods
+        .map((pe) => Number(getVal({ response, dx, pe, ou })))
+        .map((val, index) => [index, val])
+        .filter((point) => !isNaN(point[1]))
+
+    if (trend === 'constant') {
+        const comparisonPeriodValues = comparisonPeriodPoints.map(
+            (point) => point[1]
+        )
+        return (currentPeriodValue / getMean(comparisonPeriodValues)) * 100
+    }
+    // else forecast
+    return getForecastValue({
+        pointsArray: comparisonPeriodPoints,
+        forecastX: comparisonPeriods.length,
+    })
+}
+
+const calculateSection2d = ({
+    overallResponse,
+    levelOrGroupResponse,
+    mappedConfiguration,
+    metadata,
+    overallOrgUnit,
+    subOrgUnitIDs,
+    currentPeriodID,
+    comparisonPeriods,
+}) => {
+    const results = {
+        section2d: [],
+    }
+
+    for (const dx in overallResponse) {
+        // retrieve info for dx
+        const dxInfo = mappedConfiguration?.dataElementsAndIndicators?.[dx]
+        if (!overallResponse[dx][overallOrgUnit][currentPeriodID]) {
```

Comment:
🤕 I guess this is an unfortunate side effect of having the refetch trigger in the useEffect hook...because I guess there's a couple of seconds where there's `data` and it doesn't correspond to the period. I guess handling with the optional chaining should be sufficient (and we want to handle missing data anyway), although maybe refactoring to avoid this would be a non-MVP solution.

I've done `if (!overallResponse[dx]?.[overallOrgUnit]?.[currentPeriodID])` in case the data element is not present due to the group changing.